staging/src/k8s.io/csi-api/pkg/apis/csi/v1alpha1/types.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriver captures information about a Container Storage Interface (CSI)
+// volume driver deployed on the cluster.
+// CSIDriver objects are non-namespaced.
+type CSIDriver struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec `json:"spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CSIDriverList is a collection of CSIDriver objects.
+type CSIDriverList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+	// Items is the list of CSIDriver
+	Items []CSIDriver `json:"items"`
+}
+
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+	// Driver indicates the name of the CSI driver that this object refers to.
+	// This MUST be the same name returned by the CSI GetPluginName() call for
+	// that driver.
+	Driver string `json:"driver"`
+
+	// Indicates this CSI volume driver requires an attach operation (because it
+	// implements the CSI ControllerPublishVolume() method), and that Kubernetes
+	// should call attach and wait for any attach operation to complete before
+	// proceeding to mounting.
+	// If value is not specified, default is true -- meaning attach will be
+	// called.
+	// +optional
+	AttachRequired *bool `json:"attachRequired"`
+
+	// Indicates this CSI volume driver requires additional pod information
+	// (like podName, podUID, etc.) during mount operations.
+	// If this is set to true, Kubelet will pass pod information as
+	// VolumeAttributes in the CSI NodePublishVolume() calls.
+	// If value is not specified, default is false -- meaning pod information
```

Comment:
"Backwards compatibility" and "ability to detect `false` vs user did not specify" are the usual reasons for making boolean fields a pointer. Neither of those are an issue here, so consistency? The `AttachRequired` field above needs to be able to differentiate between `false` vs user did not specify, so it has to be a pointer. And all future boolean fields added here will have to be pointers for backwards compat. So consistency with the other field here and future fields added.

Reply:
My question was why this is an option at all - under what circumstances would you not want that info?  I mean, some might not need it, but why not pass it anyway?

Reply:
Currently all plugins are supposed to validate every volume attribute and error on attributes they don't recognize.  This could break plugins when we upgrade Kubernetes independently of the plugin.

Reply:
+1 to @msau42. These parameters are not part of the CSI spec. Most plugins won't use them. The plugins that need them must opt-in to this behavior.

Reply:
blech, can we fix CSI to be less brittle?

Reply:
> blech, can we fix CSI to be less brittle?

We looked in to adding workload info into CSI. The challange is that CSI is agnostic to the CO (has to work with Kubernetes, Mesos, and others). So Kubernetes specific pod information can't be encoded in to the spec. But we'd still like to be able to have CSI drivers that want this Kube specific info to be able to receive it. And this is a relatively clean way to do so (leaves CSI spec agnostic, and allows individual drivers to choose not to be).

Reply:
> It sounds like this should be a version counter that controls the format of the data passed, not a true/false field?

That's a good idea. Changed this from a boolean to a version counter: `PodInfoOnMountVersion *string` with the comment documentation capturing accepted version value.